clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -472,6 +472,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (TimeoutException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
```

Comment:
@ijuma  Actually yeah, it can even happen when updating metadata. My bad. This seems like a viable solution or we can throw a BuffereExhaustedException from the Bufferpool.allocate() method instead of TimeoutException and have a proper message of what happened. What do you think?